[SPARK-17204][CORE] Fix replicated off heap storage

mallman opened the pull request:
(Jira: https://issues.apache.org/jira/browse/SPARK-17204)

## What changes were proposed in this pull request?

There are a couple of bugs in the `BlockManager` with respect to support for replicated off-heap storage. First, the locally-stored off-heap byte buffer is disposed of when it is replicated. It should not be. Second, the replica byte buffers are stored as heap byte buffers instead of direct byte buffers even when the storage level memory mode is off-heap. This PR addresses both of these problems.

## How was this patch tested?

`BlockManagerReplicationSuite` was enhanced to fill in the coverage gaps. It now fails if either of the bugs in this PR exist.

Commit:
Fix a couple of bugs in replicated off-heap storage

Commit:
Shore-up BlockManagerReplicationSuite to identify a couple of bugs with
off-heap storage replication

mallman commented on core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:
```diff
@@ -387,12 +388,23 @@ class BlockManagerReplicationSuite extends SparkFunSuite
         val testStoreName = testStore.blockManagerId.executorId
-        assert(
-          testStore.getLocalValues(blockId).isDefined, s"$blockId was not found in $testStoreName")
-        testStore.releaseLock(blockId)
```
N.B. We no longer need the `releaseLock` call because we exhaust the iterator returned by `getLocalValues`.

SparkQA commented:
**[Test build #71022 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71022/testReport)** for PR 16499 at commit [`e49aeca`](https://github.com/apache/spark/commit/e49aeca23ff463fbd9a9cc4db99078c466bfbd56).

mallman commented on core/src/test/scala/org/apache/spark/storage/BlockManagerReplicationSuite.scala:
```diff
@@ -375,7 +375,8 @@ class BlockManagerReplicationSuite extends SparkFunSuite
       val blockId = new TestBlockId(
         "block-with-" + storageLevel.description.replace(" ", "-").toLowerCase)
-      stores(0).putSingle(blockId, new Array[Byte](blockSize), storageLevel)
+      val testValue = Array.fill[Byte](blockSize)(1)
```
Using an array of 1s instead of an array of 0s is my silly, paranoid, OCD way of adding a little extra entropy to the test. I think the chance that this change in test value will actually affect the outcome of this test is about 0%. I will revert to the original test value on request.

SparkQA commented:
**[Test build #71022 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/71022/testReport)** for PR 16499 at commit [`e49aeca`](https://github.com/apache/spark/commit/e49aeca23ff463fbd9a9cc4db99078c466bfbd56).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/71022/
Test PASSed.

mallman commented:
@rxin, can you recommend someone I reach out to for help reviewing this PR?

srowen commented:
Ideally I think @joshrosen is the person to take a look

mallman commented:
Josh, can you take a look at this when you have a chance?

rxin commented:
also cc @sameeragarwal

viirya commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1018,7 +1025,9 @@ private[spark] class BlockManager(
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.dispose()
+            if (!level.useOffHeap) {
```
Do we need to call dispose on on-head byte buffer? I think only off-heap byte buffer needs to be disposed?

mallman replied:
I think the name of the `ChunkedByteBuffer.dispose()` method is confusing. It actually only attempts to dispose a so-called "memory mapped" buffer. On-heap buffers are not memory mapped, therefore this is a no-op for them.

On the other hand, when the storage level uses off-heap memory in this context, `bytesToReplicate` is a reference to the actual off-heap memory buffer. Disposing of this buffer will erase it from the local memory store. Obviously, this is not the desired behavior. So we add the guard for off-heap memory buffers here.

As far as I can tell, there is no storage level for which `bytesToReplicate.dispose()` would actually do anything. However, technically if `bytesToReplicate` where memory-mapped but not direct, this would dispose of that memory. Would we even want that behavior? Overall, this `finally` clause is attempting to destroy the data we get from `doGetLocalBytes()`. This does not seem to be safe or correct, because we do not want to destroy the data stored locally. Therefore, we should consider getting rid of this finally clause entirely. What do you think?

sameeragarwal replied:
Similarly, aren't we deciding whether we should dispose `bytesToReplicate` based on the 'target' `StorageLevel`? Shouldn't we make that decision based on whether `bytesToReplicate` were stored off-heap or not?

mallman replied:
So maybe use `putBlockStatus.storageLevel` instead?

viirya replied:
As `StorageUtils.dispose` only cleans up a memory-mapped `ByteBuffer`, I don't think calling `bytesToReplicate.dispose()` here would be a problem.

mallman replied:
Allocating a direct byte buffer creates a `java.nio.DirectByteBuffer`, which is in turn a subclass of `java.nio.MappedByteBuffer`. So calling `dispose()` will dispose direct buffers, too.

viirya replied:
> Allocating a direct byte buffer creates a java.nio.DirectByteBuffer, which is in turn a subclass of java.nio.MappedByteBuffer. So calling dispose() will dispose direct buffers, too.

yeah, right.

If `bytesToReplicate` comes from disk store, it could be a memory-mapped byte buffer, doesn't this change may miss the change to dispose it?

mallman replied:
There does not appear to be a robust way to check if `bytesToReplicate` is a mmapped file or not. Perhaps `doGetLocalBytes` should return a tuple `(ChunkedByteBuffer, boolean)` where the second element of the tuple is `true` if and only if the buffer is a mmapped file. Thoughts?

sameeragarwal replied:
I agree, we can have `doGetLocalBytes` return an additional `boolean` that indicates whether the returned buffer is memory-mapped or not.

viirya replied:
Sounds good.

cloud-fan replied:
+1

mallman replied:
Hi guys. Sorry for the delay on the update. I started down the path that I proposed and it resulted in too many awkward changes in method signatures downstream. I don't think this is a viable step forward.

As another option, we could dispose the buffer if and only if it has a non-null `fd` field. Since that field is private, we would have to call it by reflection. I'd also include a unit test to validate that the field exists as expected to guard against internal changes in future versions of Java.

On a broader level, I wonder if callers of `ChunkedByteBuffer.dispose` method understand that it will dispose of non-memory-mapped direct buffers? The documentation of that method suggests it's only supposed to dispose of memory-mapped files in the strict sense (those actually memory mapped against a file descriptor by the OS). If other methods are accidentally calling this method on non-memory-mapped direct buffers, that suggests to me we need to push the fix to that method (or actually the StorageUtils.dispose() method). What do you think of that?

cloud-fan replied:
`StorageUtils.dispose` is a good place to put the fix.

viirya replied:
To fix it in StorageUtils.dispose() sounds good.

mallman replied:
Ok. I'll see what else is calling that method to validate that a fix there won't break something else, and I'll add a unit test to validate that calling `StorageUtils.dispose` on a direct byte buffer that isn't memory mapped doesn't actually dispose it.

viirya commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -813,7 +813,14 @@ private[spark] class BlockManager(
-          memoryStore.putBytes(blockId, size, level.memoryMode, () => bytes)
+          val memoryMode = level.memoryMode
+          memoryStore.putBytes(blockId, size, memoryMode, () => {
+            if (memoryMode == MemoryMode.OFF_HEAP) {
```
Do we need to check if `bytes` is already a direct buffer?

mallman replied:
(Actually, I think we need to check `memoryMode.useOffHeap` here.)

Assume `memoryMode.useOffHeap` is true. We have two cases to consider:

1. `bytes` is on-heap. In this case, we need to copy it into a new direct buffer, and that's what we're doing here.
2. `bytes` is off-heap. In this case, we assume that the caller upstream is managing the memory underlying `bytes`, and `bytes.copy(Platform.allocateDirectBuffer)` becomes a defensive copy. If the caller is not managing this memory, I would call that a bug in the caller's behavior.

In either case, I believe we should be calling `bytes.copy(Platform.allocateDirectBuffer)` when `memoryMode.useOffHeap` is true.

BTW, in my experience tracing this code in the debugger, `bytes` has always been an on-heap buffer.

mallman replied:
NM about checking `memoryMode.useOffHeap`. I got that confused with `StorageLevel`. There's actually only two values of `MemoryMode`: `MemoryMode.OFF_HEAP` and `MemoryMode.ON_HEAP`.

mallman replied:
Is it safe to store a ref to `bytes` if the memory is stored off-heap? If the caller changes the values in that memory or frees it, the buffer we put in the memory store will be affected. We don't want that kind of side-effect.

viirya replied:
I did a search. Looks like the buffers passed to `putBytes` (the only caller of private `doPutBytes`) across Spark are all duplicated.

mallman replied:
I'm not sure what you're suggesting I do here.

viirya replied:
I mean if all `bytes` are copy before passing in, it is no need to do another copy.

mallman replied:
So do a copy if and only if `memoryMode == MemoryMode.OFF_HEAP` and `bytes` is not direct and `bytes` is not a memory mapped file?

viirya replied:
No. I meant do we actually need to defensive copy here? All usage of `putBytes` across Spark have duplicated the byte buffer before passing it in. Is there any missing case we should do this defensive copy?

viirya replied:
Oh. nvm. That `duplicate` doesn't actually copy buffer content...

viirya replied:
But I am still wondering if we need to do copy like this here. Right, it is defensive, but as `BlockManager` is private to spark internal, and if all callers to it do not modify/release the byte buffer passed in, doesn't this defensive copy only cause performance regression?

Commit:
When storing a ChunkedByteBuffer to the memory store at the OFF_HEAP
storage level, do not copy the underlying data unless the buffer is not
direct

SparkQA commented:
**[Test build #73223 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73223/testReport)** for PR 16499 at commit [`320b548`](https://github.com/apache/spark/commit/320b54884fc086806f28bda15c10e091ed4edf72).

SparkQA commented:
**[Test build #73226 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73226/testReport)** for PR 16499 at commit [`25923f3`](https://github.com/apache/spark/commit/25923f36ce7a2e2b94ab6ae5a915ba8b1c05cd25).

mallman commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -843,7 +852,15 @@ private[spark] class BlockManager(
+          val memoryMode = level.memoryMode
+          memoryStore.putBytes(blockId, size, memoryMode, () => {
+            if (memoryMode == MemoryMode.OFF_HEAP &&
+                bytes.chunks.exists(buffer => !buffer.isDirect)) {
```
I've refined this check for copying `bytes` to skip copying when the underlying buffers are already direct.

mallman commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -317,6 +317,9 @@ private[spark] class BlockManager(
   /**
    * Put the block locally, using the given storage level.
+   *
+   * '''Important!''' Callers must not mutate or release the data buffer underlying `bytes`. Doing
```
I've explicitly documented the fact that callers must not mutate the data buffers underlying `bytes`.

SparkQA commented:
**[Test build #73223 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73223/testReport)** for PR 16499 at commit [`320b548`](https://github.com/apache/spark/commit/320b54884fc086806f28bda15c10e091ed4edf72).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #73226 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73226/testReport)** for PR 16499 at commit [`25923f3`](https://github.com/apache/spark/commit/25923f36ce7a2e2b94ab6ae5a915ba8b1c05cd25).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73223/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73226/
Test PASSed.

Commit:
Add unmap

Commit:
Clarify and expand the documentation of the
ChunkedByteBuffer.toByteBuffer method to ensure callers understand
this method does not always return a copy of the underlying data

SparkQA commented:
**[Test build #74127 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74127/testReport)** for PR 16499 at commit [`45eb006`](https://github.com/apache/spark/commit/45eb0066dc0eb4003e371735855825bcd1c91524).

mallman commented:
I looked into simply cleaning up the `StorageUtils.dispose` method to only dispose memory-mapped buffers. However, I did find legitimate uses of that method to dispose of direct/non-memory-mapped buffers. So I kept the behavior of that method as-is. Instead, I added a new method—unmap—which will dispose of memory-mapped buffers *only*, and added calls to that method where appropriate. At the end of the day, I only found one case where we specifically wanted an "unmap" behavior instead of the other broader disposal behavior. (That case being the one what was causing corruption of replicated blocks in the first place.)

I also found a new memory management bug in `BlockManager` introduced by the encryption support. In the original codebase, it disposes of a buffer unsafely. I think part of the problem is the documentation of the `ChunkedByteBuffer.toByteBuffer` method uses the word "copy" in describing what that method does. I expanded and made that method's documentation more precise to clarify that sometimes that method *does not* return a copy of the data. In those cases, it is not safe to dispose the returned buffer.

I found that there were no uses of `ByteBufferInputStream` where automatic buffer disposal was called for. Therefore, I dropped that support from that class to guard against unsafe usage. If someone _really_ wants to actually use automatic buffer disposal in `ByteBufferInputStream` they can—carefully—re-add that support. I think that that's generally unsafe. And, like I said, nothing in the codebase was using it anyway except where it was used incorrectly.

cloud-fan commented on core/src/main/scala/org/apache/spark/storage/BlockManager.scala:
```diff
@@ -1048,7 +1065,7 @@ private[spark] class BlockManager(
             replicate(blockId, bytesToReplicate, level, remoteClassTag)
           } finally {
-            bytesToReplicate.dispose()
+            bytesToReplicate.unmap()
```
one new thought: ideally we only wanna dispose bytes returned by `DiskStore`. `ChunkedByteBuffer` has a boolean flag, `disposed`. If this flag is true, `ChunkedByteBuffer.dispose` will become no-op.

When `MemoryStore` returns a `ChunkedByteBuffer`, how about we setting the `disposed` flag to true? Then we won't dispose it later.`

mallman replied:
I would to explore this further to ensure this would really work well, but I like your idea with one caveat. I think we should avoid using the `disposed` var for this purpose. Using it this way would introduce ambiguity in its meaning when `disposed` is true. In addition to its current meaning, it could also mean "this buffer is not disposed but we don't want to dispose it".

Instead, I suggest keeping the usage of `disposed` as-is and adding an additional var, e.g. `indisposable`, which makes the `dispose` method itself a no-op.

mallman replied:
@cloud-fan I explored the approach of making the `MemoryStore` return a `ChunkedByteBuffer` that cannot be disposed, however I don't think there's a clean way to safely support that behavior. In essence, if the memory manager marks a buffer as indisposable when it returns it to the block manager, then that buffer cannot be evicted later. Adding additional code to handle this other behavior correctly was looking rather messy, and I abandoned the effort.

At this point, I think that explicitly separating `unmap` and `dispose` methods is still the best way to resolve this issue.

cloud-fan replied:
yea we should separate `unmap` and `dispose`. but instead of using a hack in https://github.com/apache/spark/pull/16499/files#diff-21027f5c826cd378daaae5f7c3eea2b5R240, shall we use a just a flag `needUnmap` in `ChunkedByteBuffer `?

mallman replied:
The best I think we can expect from such a flag is a hint. The constructor of a `ChunkedByteBuffer` will not always know if the underlying byte buffers are memory mapped or not. For example, see https://github.com/apache/spark/blob/bec6b16c1900fe93def89cc5eb51cbef498196cb/core/src/main/scala/org/apache/spark/storage/BlockManager.scala#L326.

In this case, `data.nioByteBuffer()` might or might not be memory-mapped.

I still think the current patch set is the best overall amongst the other options we've considered. I can add a unit test for `StorageUtils.unmap` to ensure it works as expected (only disposing memory-mapped buffers). I can also add an `if` clause around the call to `bytesToReplicate.unmap()` to ensure this is only called when the replication storage level is off-heap. This will ensure the reflective call on the `fd` field only occurs for off-heap replication. Given that off-heap replication is currently broken, I doubt anyone will notice a performance degradation... Besides that, I suspect that network and disk IO performance will dominate the reflective method call performance.

SparkQA commented:
**[Test build #74127 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74127/testReport)** for PR 16499 at commit [`45eb006`](https://github.com/apache/spark/commit/45eb0066dc0eb4003e371735855825bcd1c91524).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74127/
Test PASSed.

cloud-fan commented:
thanks, merging to master/2.1!

@mallman can you send a new PR for 2.0? thanks!

mallman commented:
> @mallman can you send a new PR for 2.0? thanks!

Will do. Do I need to open a new JIRA ticket for that?

gatorsmile commented:
You do not need to open the new JIRA. You can still use the same JIRA number

mallman commented:
Backport PR is #17390
